### PR TITLE
Add support for multiple programming languages

### PR DESCRIPTION
**Pull request by matthieubulte**
This pull request in not intended to be merged right now as it's not quite ready. Still I decided to open it in order to have some feedback on the changes introduced in order to support multiple programming languages and the implementation for each of the new programming languages.

We also have to discuss several topics:
+ how to have a friendlier choice of the programming language
+ where to store which language to use when opening tddbin (url?)
+ should we use the compilers' feedback to highlight errors or warnings in the code? if yes, we should discuss the UX and if we want to introduce this now or later.

Please add feedbacks, and be rude if needed!

### CODE REVIEW

File: src/main/main-controller.js [hunk 1]
```diff
@@ -2,11 +2,21 @@ import React from 'react';
 import View from './main-view';
 import TestRunner from '../test-runner/runner';
 import ShortcutProcessor from '../keyboard-shortcut/shortcut-processor';
+var compiler = require('../test-runner/compiler');
```
Comment (wolframkriesing) [inline src/main/main-controller.js]: i would use `import` here too, lets get more ES6ified :)

**Comment by wolframkriesing** [comment]
The more I think about it, the more I tend towards having an external solution (microservice) that does the transpiling to es5, from whatever language.
So that we might do something like this:
```
POST toES5.tddbin.com/?from=es6
content: ....lots of es6 code....
```

input is
- source code
- language name (es6, purescript, coffeescript, typescript, clojurescript, ...)
output
- error or es5 code

looks pretty simple from the outside, and using [xdomain] we can have it be a pure client side service, i think.

[xdomain]: https://github.com/jpillora/xdomain

**Comment by matthieubulte** [comment]
Well, it seems to be quite a big one to do.

I can have a look at it tonight, but I'd rather get the multi language functionality in and then migrate to xdomain the parts that needed, namely compiling and katas.

**Comment by tddbin** [comment]
thinking of the katas too, i did update the directory structure yesterday,
so the katas-service can
provide katas in any language, see
https://github.com/tddbin/katas-service/tree/master/katas
what do you think? this way, this service can be the bucket for all kinds
of katas, independent of all the
languages tddbin supports, etc. etc.
Makes sense?

--
Kind regards / Saludos / Mit freundlichen Grüßen
Things we think are important http://www.uxebu.com/universe/
Meet uxebus at http://www.uxebu.com/about-us/events/

Wolfram Kriesing, CTO
uxebu

kriesing@uxebu.com, mobile: +49 174 300 4595

uxebu GmbH
Fraunhofer Straße 31e
86167 Augsburg

Amtsgericht Augsburg, Handelsregister HRB 28613
Directors: W. Kriesing, T. v. Klipstein

**Comment by matthieubulte** [comment]
Looks cool. Brings me to these ideas of services:

+ configuration
    + input: none
    + output: `languageConfigurations`. A language configuration is an object specifying:
        + `name`: name of the language
        + `compilerUrl`: who to send the compiled code to
        + `availableFrameworks`: an array of
            + `name`: name of the framework
            + `dependenciesUrls`: array of urls to send to the compiler

+ compiler
    + input: `language` and `sourceCode` and `dependenciesUrls`
        + `language`: name of the used language
        + `dependenciesUrls`: list of urls to download dependencies from
        + `sourceCode`: code to actually compile
    + output: `runnableCode` or `errors`

+ all-katas
    + input: none
    + output: `availableKatas`: array of url to available katas, an available kata is:
        + `name`: name of the kata
        + `url`: url to the source file
        + `language`: name of the language
        + `framework`: name of the framework

+ filesystem
    + input: `pathToFile`
    + output: `content`

Now, each of these services can (and will) talk to each other, and have different backends. But we are flexible on every choice (offline / online or github / real database)

**Comment by tddbin** [comment]
Cool ideas.
What do you think of decoupling those services as much as possible to
create as little as possible dependencies between all of them. Best would
be none at all. So that actually each of them can stand alone. Micro
services at its best ;)
 Am 09.03.2015 14:31 schrieb "matthieubulte" <notifications@github.com>:

**Comment by matthieubulte** [comment]
well, the only coupling you would have is to the `filesystem` service. The others don't have to know about each other.

I thought we could do the following:
1. Polish this branch to merge and deploy it. Then we can have some feedback and feel secured about the feature itself.
2. Rewrite the Katas using client and server-side micro services using xdomain.
3. Deploy the new katas to get some feedback.
4. Do the rest :)

Now what do you think would be the next steps?

**Comment by wolframkriesing** [comment]
Prototyping it and getting feedback is definitely the best step to do
first. So step 1) is what to do next, I agree.
Am 09.03.2015 15:49 schrieb "matthieubulte" <notifications@github.com>:

**Pull request closed without merging**